util/memory/tracker.go

```diff
@@ -39,97 +39,119 @@ import (
 // 2. Other operations of a Tracker tree is not thread-safe.
+//
+// We have two thresholds for the memory limit: soft limit and hard limit.
+// If the soft limit(0.8 * hard limit) is exceeded, we will triage the action that alleviate the speed of memory growth.
+// If the hard limit(memory-quota-query) is exceeded, we will triage the action that immediately reduce memory usage.
 type Tracker struct {
 	mu struct {
 		sync.Mutex
 		// The children memory trackers. If the Tracker is the Global Tracker, like executor.GlobalDiskUsageTracker,
 		// we wouldn't maintain its children in order to avoiding mutex contention.
 		children map[int][]*Tracker
 	}
-	actionMu struct {
-		sync.Mutex
-		actionOnExceed ActionOnExceed
-	}
-	parMu struct {
+	actionMuForHardLimit actionMu
+	actionMuForSoftLimit actionMu
+	parMu                struct {
 		sync.Mutex
 		parent *Tracker // The parent memory tracker.
 	}
 
-	label         int   // Label of this "Tracker".
-	bytesConsumed int64 // Consumed bytes.
-	bytesLimit    int64 // bytesLimit <= 0 means no limit.
-	maxConsumed   int64 // max number of bytes consumed during execution.
-	isGlobal      bool  // isGlobal indicates whether this tracker is global tracker
+	label          int   // Label of this "Tracker".
+	bytesConsumed  int64 // Consumed bytes.
+	bytesHardLimit int64 // bytesHardLimit <= 0 means no limit.
+	bytesSoftLimit int64
+	maxConsumed    int64 // max number of bytes consumed during execution.
+	isGlobal       bool  // isGlobal indicates whether this tracker is global tracker
+}
+
+type actionMu struct {
+	sync.Mutex
+	actionOnExceed ActionOnExceed
 }
 
+// softScale means the scale of soft limit to hard limit.
+const softScale = 0.8
+
 // NewTracker creates a memory tracker.
 //	1. "label" is the label used in the usage string.
-//	2. "bytesLimit <= 0" means no limit.
+//	2. "bytesHardLimit <= 0" means no limit.
 // For the common tracker, isGlobal is default as false
 func NewTracker(label int, bytesLimit int64) *Tracker {
 	t := &Tracker{
-		label:      label,
-		bytesLimit: bytesLimit,
+		label:          label,
+		bytesHardLimit: bytesLimit,
 	}
-	t.actionMu.actionOnExceed = &LogOnExceed{}
+	t.bytesSoftLimit = int64(float64(bytesLimit) * softScale)
+	t.actionMuForHardLimit.actionOnExceed = &LogOnExceed{}
 	t.isGlobal = false
 	return t
 }
 
 // NewGlobalTracker creates a global tracker, its isGlobal is default as true
 func NewGlobalTracker(label int, bytesLimit int64) *Tracker {
 	t := &Tracker{
-		label:      label,
-		bytesLimit: bytesLimit,
+		label:          label,
+		bytesHardLimit: bytesLimit,
 	}
-	t.actionMu.actionOnExceed = &LogOnExceed{}
+	t.bytesSoftLimit = int64(float64(bytesLimit) * softScale)
+	t.actionMuForHardLimit.actionOnExceed = &LogOnExceed{}
 	t.isGlobal = true
 	return t
 }
 
 // CheckBytesLimit check whether the bytes limit of the tracker is equal to a value.
 // Only used in test.
 func (t *Tracker) CheckBytesLimit(val int64) bool {
```

Comment:
Should rename the Checker/Setter/Getter?

Reply:
I don’t think it’s necessary. I think hard/soft limit is just a concept inside tracker.